### DISCUSSION TITLE
Running `aqt list-qt` on a web server

### DISCUSSION THREAD

**Discussion by ddalcino** [miurahr/aqtinstall#426] [Show and tell]
I think that since `aqtinstall` is often used in a remote CI environment, many of our users never even get the chance to interact with the command line interface directly, which means that they might not have the opportunity to run `aqt list-*`. They could install `aqt` locally, just to run `aqt list-qt`, so they can figure out how to configure their CI environment, but that's a lot of time and effort they might rather spend in some other way. What if they could interact with `aqt list-*` from a website instead?

This website defines a simple REST api that makes `aqt list-*` accessible from the web. It also provides a simple GUI client, written in React/Typescript, that allows you to select your host, target, version, arch, modules, and tools from the available options, and it tells you how to use `aqt install-*` or the `jurplel/install-qt-action` to install whatever you want. The website is not done yet, and there is lots of potential for improvement. You can visit the website here: https://aqt-list-app.herokuapp.com/

**Comment by pzhlkj6612** [comment]
> You can visit the website here: https://aqt-list-app.herokuapp.com/

Ohhhh, nice work bro!

Question: are you running an instance of aqtinstall on heroku?

**Reply by ddalcino** [reply]
Thanks!

It’s a Python Flask app that uses aqtinstall 2.0.0 as a dependency. The REST endpoints call aqt.MetadataFactory directly, and return a json object with the output. It is hosted on heroku.

**Reply by pzhlkj6612** [reply]
OK, I got it. Now I'm thinking of:

How about making a JavaScript app running on user's browser? It will allow the client to fetch data directly from https://download.qt.io/ or even other mirror sites. I think this might be another way to achieve it, but more work definitely needs to be done (project diverged: aqtinstall[<sup>**TS**</sup>](https://www.typescriptlang.org/)).

**Reply by ddalcino** [reply]
Yes; if you could port `aqt list-qt` to javascript, you could eliminate the Heroku instance. This would probably be an improvement. However, that seemed like too much work for a quick prototype like this.

When I started this, I tried to use one of the "Python-JS" transpilers to transform aqt directly into JS, but I didn't get very far. I also tried brython, which is a python interpreter that runs in the browser. I never figured out how to get aqt to run that way.

Running `aqt list-qt` in a Heroku instance will be problematic if not enough people use it (Heroku could turn off the server), or if too many people use it. I have a free account, so there are usage limits.

The Heroku instance does offer some potential advantages though. Right now, I have some limited caching set up in the Flask app that caches responses from https://download.qt.io/. From my location, any access to an Updates.xml file at https://download.qt.io/ takes nearly 1 second; responses from Heroku are a couple of milliseconds. I haven't done this yet, but I think you can configure nginx to cache the responses to all the REST api calls, which could make the REST api significantly faster than any solution that queries https://download.qt.io/ directly. 

**Edit:** I didn't realize that Heroku just shuts down the server a *very* short time after you stop using it. It takes almost a minute (I think) to spin the server back up. A JS or TS port of list-qt would be very helpful here.

**Reply by ddalcino** [reply]
> How about making a JavaScript app running on user's browser? It will allow the client to fetch data directly from https://download.qt.io/ or even other mirror sites.

There’s no way to download anything from https://download.qt.io/ or the mirror sites from within a browser because the server CORS policy forbids it. You would have to fetch the data from some other source.
